### PR TITLE
Fix ambiguous prerelease retries

### LINKED ISSUE #4646: Release: argument-less prerelease retry can accidentally propose stable promotion
## Problem

An argument-less `bundle exec rake release` can silently turn an interrupted prerelease retry into a stable release proposal.

Concrete rc.10 incident:

1. The first `release[17.0.0.rc.10]` run pushed version-bump commit `d09328988dd5caa7e358f262174141d0a84b7f94` and tag `v17.0.0.rc.10`, then stopped before publishing npm packages, RubyGems, or the GitHub release.
2. The operator retried with `RELEASE_CI_STATUS_OVERRIDE=true bundle exec rake release` (no version argument).
3. `resolve_version_input` found the changelog version equal to the current version (`17.0.0.rc.10`) and fell back to `patch`.
4. `compute_target_gem_version` correctly implements gem-release semantics where a patch bump from a prerelease strips the suffix, producing stable `17.0.0`.
5. The version-policy check found no `17.0.0` changelog section but only logged that it was skipping consistency validation.
6. The confirmation prompt warned `Changelog: ✗ MISSING` but still allowed `Proceed with release? [y/N]`.

This creates a high-risk path to accidentally tag/publish stable `17.0.0`—including npm `latest`—while intending only to resume an incomplete rc.10 publication.

## Desired behavior

An argument-less release from a prerelease checkout must never silently infer stable promotion.

Recommended behavior:

- If current version is a prerelease and the latest changelog version is equal, detect an idempotent retry candidate and either:
  - retry that exact prerelease when tag/registry state proves a partial release, or
  - abort with a copy-paste explicit command such as `bundle exec rake "release[17.0.0.rc.10]"`.
- Stable promotion from an RC must require an explicit stable version (or the dedicated release-finish flow) plus a matching stable changelog section.
- A non-dry-run stable release with no matching changelog section should fail closed rather than merely warn and accept `y`.
- Preserve the normal argument-less patch behavior for already-stable releases if that remains intentional.

## Acceptance criteria

- [ ] `rake release` from current `17.0.0.rc.10` with changelog latest `17.0.0.rc.10` cannot propose `17.0.0` without an explicit promotion request.
- [ ] A partial prerelease state (version commit/tag exists, some or all registry artifacts missing) prints the exact idempotent retry command.
- [ ] Explicit `release[17.0.0.rc.10]` continues to use the existing tag-at-HEAD/idempotent publication retry path.
- [ ] Explicit stable promotion requires a `17.0.0` changelog section and the release-branch promotion invariants.
- [ ] Missing stable changelog content blocks before confirmation/tag/publication.
- [ ] Tests cover: interrupted prerelease before tag, after tag/before registries, partially published registries, completed prerelease, explicit stable promotion, and normal stable patch bump.
- [ ] Documentation warns operators never to drop the version argument when resuming a partial release.

## Current recovery

For the rc.10 incident, abort the accidental `17.0.0` prompt and retry only after exact-head CI completes:

```bash
RELEASE_CI_EVALUATE_HEAD=true bundle exec rake "release[17.0.0.rc.10]"
```

Do not use `RELEASE_CI_STATUS_OVERRIDE` while exact-head checks are pending.

### PR DESCRIPTION
**Pull request by justin808**
## Summary

Argument-less release retries now fail closed when the command starts from a prerelease checkout and the changelog is missing, stale, stable, on another release line, or still points at the same prerelease. The release task preserves the starting prerelease version across its pull, so an interrupted remote stable-version bump cannot erase that ambiguity before the safety check runs.

The abort happens before npm/GitHub authentication checks, release-tag lookup, or registry lookup and prints the exact explicit prerelease retry command. Stable releases and stable dry runs also require a non-empty matching changelog section before operator confirmation, tagging, or publication. A clearly newer prerelease on the same release line remains available. An already-stable argument-less checkout may still derive a patch candidate, but the stable changelog gate blocks publication until that candidate has matching non-empty evidence.

Prerelease dry runs now report missing changelog evidence without prompting. The release guide and task help distinguish stable aborts from prerelease warnings and scope post-release changelog repair to prereleases or historical releases.

## Decision log

- Preserve the version observed before `git pull --rebase` only when the invocation is argument-less and starts from a prerelease; stable checkouts continue resolving against the post-pull version.
- Permit argument-less automation only for a clearly newer prerelease on the same base version.
- Require explicit stable promotion plus a non-empty matching changelog section; release-gate overrides do not bypass this prerequisite.
- Keep explicit prerelease retries on their existing tag-at-HEAD and idempotent registry-publication paths.

## Validation

Validated at exact head `efe09eb6c71717b79217e1634455eb2c7f943a19`:

- `bundle exec rspec spec/react_on_rails/release_rake_helpers_spec.rb` — 252 examples, 0 failures.
- `.agents/bin/validate --fast` — docs sidebar, full OSS and benchmark RuboCop, package build, ESLint, Prettier, and TypeScript passed; its long gem run had reached 847 examples with 0 failures when it was stopped because `main` advanced and invalidated that head.
- After rebasing onto current `origin/main`, the focused release suite, targeted RuboCop, docs sidebar, and diff checks passed again.
- Isolated configuration regression check on both untouched `origin/main` and the branch — 52 examples, 0 failures on each.
- Focused documentation sidebar/link checks and `git diff --check` — passed.
- Independent final-head GPT-5.6 Sol/xhigh review — no findings. Earlier findings about pull-time version advancement, dry-run gating, composed stable fallback behavior, and stale operator guidance were verified and fixed before this clean pass; one claimed clean-worktree ordering regression was rejected after direct code verification showed the guard still runs before pull.

## Release and workflow notes

- Release tracker mode: development; target phase: beta.
- Confidence: high, based on fail-closed regressions, full helper-path coverage, clean independent review, and current-head local validation.
- Churn: the operator documentation was based on current `main` after #4650 merged; no overlapping editor paths remain.
- GitHub Actions: no semantic workflow or action changes, so no post-merge exercise issue is required.
- Changelog: `not_user_visible`; this changes maintainer release tooling and its contributor documentation, so no product changelog entry is added.

## Pull Request checklist

- [x] Add/update test to cover these changes
- [x] Update documentation
- [x] ~Update CHANGELOG file~ — not user-visible release tooling

Fixes #4646.

---

[Compound Engineering](https://github.com/EveryInc/compound-engineering-plugin)
Codex

## Summary by CodeRabbit

* **Release Safety**
  * Safer `rake release` version resolution for interrupted prerelease retries, preventing unsafe stable promotion when no explicit version is provided.
  * Stable releases now require a matching, non-empty `CHANGELOG.md` section before confirmation/tagging/publishing (including in dry-run).
  * Refined “auto” version behavior to correctly handle newer/stale changelog entries for prereleases vs stable.
* **Documentation**
  * Updated retry-safety guidance and clarified how argument-less `rake release` behaves.
* **Tests**
  * Added/expanded coverage for changelog gating, dry-run prompting behavior, and revised version inference/retry scenarios.

### CODE REVIEW

**Comment by justin808** [comment]
+ci-status

**Comment by justin808** [comment]
+ci-run-hosted

File: rakelib/release.rake [hunk 1]
```diff
@@ -2248,6 +2255,16 @@ def confirm_release!(version:, monorepo_root:)
   changelog_path = File.join(monorepo_root, "CHANGELOG.md")
   has_changelog = extract_changelog_section(changelog_path:, version:)
 
+  if !has_changelog && !release_prerelease_version?(version)
```
[the remark this answers is not in the document]
Comment (justin808) [reply rakelib/release.rake:2265]: Fixed in d8b60dc3e. The stable changelog gate now runs for dry runs as well as live releases, while dry runs still skip the interactive confirmation. Added regression coverage for stable dry-run failure and prerelease dry-run non-prompt behavior; the focused release helper suite passes 251 examples.

**Comment by justin808** [comment]
## Address-review summary

Scan scope: full PR review history through final-head push d8b60dc3e.

### Mattered
- Greptile P2: stable-release dry runs bypassed the changelog gate. Addressed in d8b60dc3e, covered by stable/prerelease dry-run regression tests, replied to, and resolved.

### Skipped
- Greptile summary duplicate: same dry-run finding as the resolved inline thread.
- Claude review: clean, with no actionable findings.
- CodeRabbit progress/status messages: status-only; the current-head review remains part of the merge gate.

Deferred-work tracking: none; no review work was deferred.

Next default scan starts after this comment. Say `check all reviews` to rescan the full PR.

**Comment by justin808** [comment]
+ci-status

**Comment by justin808** [comment]
+ci-run-hosted

File: internal/contributor-info/releasing.md [hunk 2]
```diff
@@ -88,12 +88,19 @@ RELEASE_VERSION_POLICY_OVERRIDE=true bundle exec rake "release[16.2.0]"
 bundle exec rake "release[16.2.0,false,true]"
 ```
 
+> **Retry safety:** Never drop the version argument when resuming an interrupted release. Retry the
+> exact prerelease version, for example `bundle exec rake "release[17.0.0.rc.10]"`. From a prerelease
+> checkout, an argument-less release fails closed unless the changelog advances the same release line
+> to a newer prerelease. Stable promotion must use an explicit stable version and a matching non-empty
+> changelog section.
```
[the remark this answers is not in the document]
Comment (justin808) [reply internal/contributor-info/releasing.md:98]: Fixed in efe09eb6c. The release guide now distinguishes stable aborts from prerelease warnings throughout, scopes post-release changelog repair to prerelease/historical releases, and aligns the task help. I also added prerelease dry-run warning behavior plus composed stable patch-candidate/changelog-gate coverage. The focused helper suite passes 252 examples.

**Comment by justin808** [comment]
## Address-review summary

Scan scope: review feedback after the previous summary through final-head push efe09eb6c.

### Mattered
- Claude inline documentation mismatch: addressed in efe09eb6c, replied to, and resolved. Stable aborts, prerelease warnings, and post-release repair scope are now consistent throughout the release guide.
- Claude task-description mismatch: addressed in efe09eb6c. Argument-less stable resolution is documented as deriving a patch candidate that still cannot pass the stable changelog gate without matching non-empty evidence.
- Independent local review found missing composed/dry-run coverage: addressed in efe09eb6c with a prerelease dry-run warning and a composed stable patch-candidate/changelog-gate regression test.

### Optional
- Claude's stale pre-pull retry-hint edge case was declined: the path remains fail-closed, and threading post-pull state into the preserved pre-pull ambiguity message would expand behavior after the final-candidate gate. No follow-up tracking is warranted.

### Skipped
- One independent local P1 ordering claim was rejected as factually incorrect after direct verification: `uncommitted_changes?` still runs before `with_release_checkout` and `git pull --rebase`.

Deferred-work tracking: none; no valuable work was deferred.

Next default scan starts after this comment. Say `check all reviews` to rescan the full PR.

**Comment by justin808** [comment]
+ci-status

**Comment by justin808** [comment]
+ci-run-hosted

File: rakelib/release.rake [hunk 3]
```diff
@@ -2911,17 +2981,23 @@ task :release, %i[version dry_run override_version_policy override_ci_status] do
...
     sh_in_dir_for_release(release_root, "git pull --rebase") unless is_dry_run
 
-    version_input = resolve_version_input(args_hash.fetch(:version, ""), release_root)
-    validate_requested_version_input!(version_input)
+    version_input = resolve_release_version_before_auth!(
+      version_input: args_hash.fetch(:version, ""),
+      monorepo_root: release_root,
+      dry_run: is_dry_run,
+      current_version: argumentless_starting_prerelease_version
+    )
```
[the remark this answers is not in the document]
Comment (justin808) [reply rakelib/release.rake:3000]: Confirmed intentional. Version resolution must run inside the refreshed release checkout so the ambiguity decision sees the post-pull tree while separately preserving the starting prerelease. This lets the safety abort happen before npm/GitHub authentication; the existing clean-worktree check still runs before the pull. No code change is needed.

File: rakelib/release.rake [hunk 4]
```diff
@@ -2388,12 +2414,52 @@ def version_tagged?(monorepo_root, version)
...
+    Refusing to infer stable #{stable_version} from an argument-less prerelease retry.
+
+    To resume #{current_version} publication, run:
+      bundle exec rake "release[#{current_version}]"
+
+    To prepare the final #{stable_version} release, add a non-empty CHANGELOG.md section for #{stable_version},
+    complete the final-release gates, then run:
+      bundle exec rake "release[#{stable_version}]"
+  ERROR
+end
```
[the remark this answers is not in the document]
Comment (justin808) [reply rakelib/release.rake:2444]: [auto-deferred]
This narrow retry-hint edge remains fail-closed; changing it would expand behavior after the final-candidate gate, so no code change or follow-up tracking is warranted.

**Comment by justin808** [comment]
## Final review disposition

Current head: `efe09eb6c71717b79217e1634455eb2c7f943a19`.

- Claude confirmed no security or correctness blockers.
- The auth-after-refresh ordering note was confirmed intentional: the ambiguity gate evaluates the refreshed tree before authentication while clean-worktree validation remains before pull.
- The stale pre-pull retry-hint edge was auto-deferred because it remains fail-closed and changing it after the final-candidate gate would expand behavior; no follow-up tracking is warranted.
- Both threads were replied to and resolved.
- Exact-head hosted CI, security preflight, and strict merge ledger are clean.
